### PR TITLE
Fix NoMethodError in kerberos/get_ticket by properly decoding ASN.1 OctetString in certificate SAN parsing for ticket reuqest --> "#20427"

### PR DESCRIPTION
**Pull request by Wopseeion**
This pull request addresses a bug where the auxiliary/admin/kerberos/get_ticket module fails with a NoMethodError because the code attempts to iterate over an OpenSSL::ASN1::OctetString object directly during certificate Subject Alternative Name (SAN) parsing.

Root Cause:
The extract_user_and_realm method in lib/msf/core/exploit/remote/kerberos/client/pkinit.rb was calling .each on an OpenSSL::ASN1::OctetString, which is not iterable. The SAN extension data must be explicitly decoded from the ASN.1 OctetString’s binary .value before iterating.

Fix Details:
1) Updated extract_user_and_realm to detect when the SAN ASN.1 value is an OctetString and decode its .value properly before    iteration.
2) Added robust handling to ensure all SAN entries (both OtherName and dNSName) are processed without errors.
3) Updated and improved RSpec tests to use real in-memory X.509 certificates with properly encoded SAN extensions—replacing mocks that caused tests to fail.
4) Ensured all edge cases and override scenarios are covered and tested.

Impact:
1) Fixes crashing issue during certificate parsing without requiring any external KDC.
2) Improves module stability and adherence to ASN.1 standards.
3) Enables valid use of certificates in the get_ticket auxiliary module, enhancing Kerberos PKINIT support.

Testing:
1) Ran comprehensive, fixed RSpec tests verifying typical and edge-case certificates, ensuring no regressions.
2)Confirmed module no longer raises NoMethodError when loading problematic PKCS#12 certificates.

### CODE REVIEW

File: external/source/exploits/drunkpotato/RogueWinRM exe/main.c
Comment (smcintyre-r7) [inline external/source/exploits/drunkpotato/RogueWinRM exe/main.c:1]: The removal of this file doesn't seem related to the issue so I'm guessing it was a mistake. Would you mind restoring it?

Comment (Wopseeion) [reply external/source/exploits/drunkpotato/RogueWinRM exe/main.c:1]: Oh yes it was, thank you for bringing it up ..i'll fix that part and update it

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
I got around to testing this and it kind of fixes the error.

You'll notice in line 4 though that the username is still an ASN1Data instance. 
```
msf auxiliary(admin/kerberos/get_ticket) > run
[*] Running module against 192.168.159.10
[!] Warning: Provided principal and realm (DC$@msflab.local) do not match entries in certificate:
[!]   * #<OpenSSL::ASN1::ASN1Data:0x00007f9119928c70>@
[*] 192.168.159.10:88 - Getting TGT for DC$@msflab.local
[-] Auxiliary aborted due to failure: unknown: Kerberos Error - KDC_ERR_CLIENT_NAME_MISMATCH (75) - PKINIT - KDC_ERR_CLIENT_NAME_MISMATCH
[*] Auxiliary module execution completed
msf auxiliary(admin/kerberos/get_ticket) >
```

You should be able to fix that by applying this patch to correct the assignment to `user`. My guess is we found an edge case and this never worked because `san_entry` is treated as an ASN1Data instance for the two lines directly before it.

```diff
diff --git a/lib/msf/core/exploit/remote/kerberos/client/pkinit.rb b/lib/msf/core/exploit/remote/kerberos/client/pkinit.rb
index f6e1ce1a2a..cdc84281ee 100644
--- a/lib/msf/core/exploit/remote/kerberos/client/pkinit.rb
+++ b/lib/msf/core/exploit/remote/kerberos/client/pkinit.rb
@@ -95,7 +95,7 @@ module Msf
                   elsif san_entry.tag == 2 # dNSName
                     parts = san_entry.value.split('.')
                     if parts.length == 1
-                      user = san_entry
+                      user = san_entry.value
                       domain = ''
                     else
                       user = parts[0] + '$'
```

Applying that change also fixes this error which would occur when `DOMAIN` and `USERNAME` are not set.

```
msf auxiliary(admin/kerberos/get_ticket) > run
[*] Running module against 192.168.159.10
[*] 192.168.159.10:88 - Getting TGT for #<OpenSSL::ASN1::ASN1Data:0x00007fc5df11b2d0>@
[-] Auxiliary failed: NoMethodError undefined method `force_encoding' for an instance of OpenSSL::ASN1::ASN1Data
[-] Call stack:
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/kerberos/client.rb:184:in `send_request_tgt_pkinit'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb:606:in `authenticate_via_kdc'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb:384:in `request_tgt_only'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/admin/kerberos/get_ticket.rb:200:in `action_get_tgt'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/admin/kerberos/get_ticket.rb:145:in `run'
[*] Auxiliary module execution completed
msf auxiliary(admin/kerberos/get_ticket) >
```

Please apply that patch and restore the main.c file. When you restore the main.c file, please rebase your changes so the commit which removed it is not in the history. With that in place, I can get this landed. Let me know if you have any questions.

**Comment by Wopseeion** [comment]
Alright i'll get it fixed

**Comment by Wopseeion** [comment]
I've done the changes ,so do i create a new pull request ?? or do i edit the existing one ? 
if i've to edit the existing one can you guide me how to?

**Comment by smcintyre-r7** [comment]
You'd push the changes up to your `problem-fi` branch which would add them to this PR.

**Comment by Wopseeion** [comment]
I've done it Can you check it and let me know ..I'm not sure if i merged my branch properly or not
in the "final" commit

**Review by smcintyre-r7 — APPROVED** [review]
I rebased your PR to undo all the superfluous changes and squash it all down to one commit. Once the tests pass, I'll get this landed.

Tested and ensure it fixed the original issue and that a certificate that used to work, continues to work.

```
msf auxiliary(admin/kerberos/get_ticket) > run
[*] Running module against 192.168.159.10
[*] 192.168.159.10:88 - Getting TGT for smcintyre@msflab.local
[+] 192.168.159.10:88 - Received a valid TGT-Response
[*] 192.168.159.10:88 - TGT MIT Credential Cache ticket saved to /home/smcintyre/.msf4/loot/20250804174617_default_192.168.159.10_mit.kerberos.cca_017319.bin
[*] Auxiliary module execution completed
msf auxiliary(admin/kerberos/get_ticket) > show options 

Module options (auxiliary/admin/kerberos/get_ticket):

   Name           Current Setting                                                                            Required  Description
   ----           ---------------                                                                            --------  -----------
   AES_KEY                                                                                                   no        The AES key to use for Kerberos authentication in hex string. Supported keys: 128 or 256 bits
   CERT_FILE      /home/smcintyre/.msf4/loot/20250804174358_default_192.168.159.10_windows.ad.cs_919541.pfx  no        The PKCS12 (.pfx) certificate file to authenticate with
   CERT_PASSWORD                                                                                             no        The certificate file's password
   DOMAIN                                                                                                    no        The Fully Qualified Domain Name (FQDN). Ex: mydomain.local
   NTHASH                                                                                                    no        The NT hash in hex string. Server must support RC4
   PASSWORD                                                                                                  no        The domain user's password
   RHOSTS         192.168.159.10                                                                             yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-metasploit.html
   RPORT          88                                                                                         yes       The target port
   Timeout        10                                                                                         yes       The TCP timeout to establish Kerberos connection and read data
   USERNAME                                                                                                  no        The domain user


   When ACTION is GET_TGS:

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   IMPERSONATE                   no        The user on whose behalf a TGS is requested (it will use S4U2Self/S4U2Proxy to request the ticket)
   Krb5Ccname                    no        The Kerberos TGT to use when requesting the service ticket. If unset, the database will be checked
   SPN                           no        The Service Principal Name, format is service_name/FQDN. Ex: cifs/dc01.mydomain.local


Auxiliary action:

   Name     Description
   ----     -----------
   GET_TGT  Request a Ticket-Granting-Ticket (TGT)



View the full module info with the info, or info -d command.

msf auxiliary(admin/kerberos/get_ticket) > set CERT_FILE bad.pfx
CERT_FILE => bad.pfx
msf auxiliary(admin/kerberos/get_ticket) > run
[*] Running module against 192.168.159.10
[*] 192.168.159.10:88 - Getting TGT for dc$@
[-] Auxiliary aborted due to failure: unknown: Kerberos Error - KDC_ERR_WRONG_REALM (68) - Wrong Realm / domain
[*] Auxiliary module execution completed
msf auxiliary(admin/kerberos/get_ticket) >
```

**Comment by smcintyre-r7** [comment]
# Release Notes
This fixes an ASN1 parsing error `auxiliary/admin/kerberos/get_ticket` that would occur when using PKINIT authentication with certain certificates.

**Comment by Wopseeion** [comment]
Alright @smcintyre-r7 , it was great solving this issue , if you've any more such issues just ping me over their, i'll look around
